Make `iframe` containing SVG behave as every other `iframe`

Ahmad-S792 opened the pull request:
#### b0064989cb96a615ee3ee6a756865b5d95aa0cfc
<pre>
Make `iframe` containing SVG behave as every other `iframe`

<a href="https://bugs.webkit.org/show_bug.cgi?id=266702">https://bugs.webkit.org/show_bug.cgi?id=266702</a>

Reviewed by Tim Nguyen.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Inspired by: <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=172467

When an SVG is contained inside an `iframe`, its dimensions should not
be part of size calculations of the `iframe`. Neither should the SVG&apos;s
own width and height be ignored when sizing the `svg` inside the
`iframe`.

* Source/WebCore/rendering/RenderWidget.cpp:
(RenderWidget::embeddedContentBox): Extended check for &apos;embeddedObject&apos;
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(LegacyRenderSVGRoot::isEmbeddedThroughFrameContainingSVGDocument): Aligned with LBSE
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/svg-embedded-sizing/svg-in-iframe-percentage-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/svg-embedded-sizing/svg-in-iframe-fixed-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/svg-embedded-sizing/svg-in-iframe-auto-expected.txt: Rebaselined
* LayoutTests/platform/glib/svg/in-html/by-reference-expected.txt: Rebaselined
* LayoutTests/svg/in-html/by-reference-expected.txt: Ditto
* LayoutTests/platform/mac/svg/in-html/by-reference-expected.txt: Ditto

Canonical link: <a href="https://commits.webkit.org/272503@main">https://commits.webkit.org/272503@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e3c6893cd77e79db1d88db4b6432c21f5142c5b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31846 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10539 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33588 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34349 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28845 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32641 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12899 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7779 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28433 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32209 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8895 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28440 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7682 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7859 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28353 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35696 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28958 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28806 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33964 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7950 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5939 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31823 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9602 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7459 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8619 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8471 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderReplaced.h:
```diff
@@ -74,10 +74,11 @@ class RenderReplaced : public RenderBox {
 
     void willBeDestroyed() override;
 
+    virtual RenderBox* embeddedContentBox() const { return 0; }
```
I'm not a huge fan of way that this bug is fixed by moving which class implements `embeddedContentBox()`. I'd prefer to see an explicit check for "this is an iframe, not an embed" or something.

Ahmad-S792 replied:
@smfr - Just to understand better, it is to create new function which specifically deals with 'iframe' and then leverage it? Also, can we do this as follow-up since it is interop bug and fails 300 WPT tests?

nt1m replied:
I think Simon is suggesting that you leave the method on `RenderWidget` and `RenderReplaced` untouched, then add a type check in `RenderWidget::embeddedContentBox()`.

So something like:

```
RenderBox* RenderWidget::embeddedContentBox() const
{
    if (!is<RenderEmbeddedObject>(this) || !is<LocalFrameView>(widget()))
        return nullptr;
    return downcast<LocalFrameView>(*widget()).embeddedContentBox();
}
```

Ahmad-S792 replied:
Above fixes this as well (checked locally) and I also manage to not do any changes in `RenderReplaced.h`.

nt1m commented on Source/WebCore/rendering/RenderReplaced.h:
```diff
@@ -74,10 +74,11 @@ class RenderReplaced : public RenderBox {
 private:
     LayoutUnit computeConstrainedLogicalWidth(ShouldComputePreferred) const;
 
-    virtual RenderBox* embeddedContentBox() const { return 0; }
```
Why did you make this method public? I don't see anything in the patch that requires it to become public.

Ahmad-S792 replied:
It is protected and not public. Without this, it leads to compiler error.

nt1m replied:
Unclear to me this needs to be protected, I don't see any new calls to `embeddedContentBox()` here.

nt1m replied:
Oh I see, RenderWidget calls embeddedContentBox but you're moving it away from RenderWidget to its subclass.

Ahmad-S792 committed:
Make `iframe` containing SVG behave as every other `iframe`

https://bugs.webkit.org/show_bug.cgi?id=266702

Reviewed by Tim Nguyen.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Inspired by: https://src.chromium.org/viewvc/blink?view=revision&revision=172467

When an SVG is contained inside an `iframe`, its dimensions should not
be part of size calculations of the `iframe`. Neither should the SVG's
own width and height be ignored when sizing the `svg` inside the
`iframe`.

* Source/WebCore/rendering/RenderWidget.cpp:
(RenderWidget::embeddedContentBox): Extended check for 'embeddedObject'
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(LegacyRenderSVGRoot::isEmbeddedThroughFrameContainingSVGDocument): Aligned with LBSE
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/svg-embedded-sizing/svg-in-iframe-percentage-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/svg-embedded-sizing/svg-in-iframe-fixed-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/svg-embedded-sizing/svg-in-iframe-auto-expected.txt: Rebaselined
* LayoutTests/platform/glib/svg/in-html/by-reference-expected.txt: Rebaselined
* LayoutTests/svg/in-html/by-reference-expected.txt: Ditto
* LayoutTests/platform/mac/svg/in-html/by-reference-expected.txt: Ditto

Canonical link: https://commits.webkit.org/272503@main
